06x-new-kafka-roller.md

```diff
@@ -0,0 +1,220 @@
+
+5. Restart nodes in `RESTART_FIRST` category either one by one in the following order unless all nodes are combined
+and are in `NOT_RUNNING` state:
+   - Pure controller nodes
+   - Combined nodes
+   - Broker only nodes
+
+   If all controllers are combined and are in `NOT_RUNNING` state, restart all nodes in parallel and wait for them to have `SERVING`. Explained more in detail below.
+
+   Wait until the restarted node to have `SERVING` within `postOperationalTimeoutMs`. If the timeout is reached and the node's `numAttempts` is greater than or equal to `maxAttempts`, throw `TimeoutException`. Otherwise increment node's `numAttempts` and restart from step 3. 
+
+6. Further refine the broker nodes in `MAYBE_RECONFIGURE` group:
+   - Describe Kafka configurations for each node via Admin API and compare them against the desired configurations. This is essentially the same mechanism we use today for the current KafkaRoller.
+   - If a node has configuration changes and they can be dynamically updated, add the node into another group called `RECONFIGURE`.
+   - If a node has configuration changes but they cannot be dynamically updated, add nodes into the `RESTART` group.
+   - If a node has no configuration changes, put the node into the `NOP` group.
+
+7. Reconfigure each node in `RECONFIGURE` group:
+   - If `numReconfigs` of a node is greater than the configured `maxReconfigs`, add a restart reason to its context. Otherwise continue.
+   - Send `incrementalAlterConfig` request with its config updates.
+   - Transitions the node's state to `RECONFIGURED` and increment its `numReconfigs`.
+   - Wait for each node that got configurations updated until they have `SERVING` within the `postOperationalTimeoutMs`. 
+   - If the `postOperationalTimeoutMs` is reached, restart from step 3.
+
+8. If at this point, the `RESTART` group is empty, the reconciliation will be completed successfully.
+
+9. Otherwise, batch nodes in `RESTART` group and get the next batch to restart:
+   - Further categorize nodes based on their roles so that the following restart order can be enforced:
+       1. `NON_ACTIVE_PURE_CONTROLLER` - Pure controller that is not the active controller
+       2. `ACTIVE_PURE_CONTROLLER` - Pure controller that is the active controller (the quorum leader)
+       3. `BROKER_AND_NOT_ACTIVE_CONTROLLER` - Node that is at least a broker but also might be a controller (combined) and is not the active controller
+       4. `BROKER_AND_ACTIVE_CONTROLLER` - Combined node that is the active controller (the quorum leader)
+
+       The returned batch will contain only one node if it is not `BROKER_AND_NOT_ACTIVE_CONTROLLER` group, so that controllers are restarted one at a time.
```

Comment:
```suggestion
       The returned batch will contain only one node for all groups other than `BROKER_AND_NOT_ACTIVE_CONTROLLER`, so that controllers are restarted one at a time.
```

But this doesn't actually guarantee controllers are restarted one at a time, does it? We need a `PURE_BROKER` group to make that guarantee.

Reply:
Adding another group for just brokers make sense!